? target/linux/image/ar7/src/entry.c
? target/linux/image/ar7/src/loader.h
? target/linux/image/ar7/src/uart16550.c
? target/linux/image/ar7/src/uart16550.h
? target/linux/linux-2.4/patches/ar7/009-diag_driver.patch
Index: target/linux/image/ar7/Makefile
===================================================================
RCS file: /openwrt/openwrt/target/linux/image/ar7/Makefile,v
retrieving revision 1.11
diff -u -b -B -u -r1.11 Makefile
--- target/linux/image/ar7/Makefile	1 Sep 2005 20:08:18 -0000	1.11
+++ target/linux/image/ar7/Makefile	2 Sep 2005 18:44:58 -0000
@@ -25,12 +25,18 @@
 	sed -e 's/@@OUTPUT_FORMAT@@/$(OUTPUT_FORMAT)/' \
 	    -e 's/@@LOADADDR@@/$(LOADADDR)/' <$< >$@
 
+$(PKG_BUILD_DIR)/entry.o: src/entry.c
+	$(TARGET_CC) $(CFLAGS) -c -o $@ $<
+
 $(PKG_BUILD_DIR)/LzmaDecode.o: src/LzmaDecode.c
 	$(TARGET_CC) $(CFLAGS) -c -o $@ $<
 
 $(PKG_BUILD_DIR)/loader.o: src/loader.c
 	$(TARGET_CC) $(CFLAGS) -c -o $@ $<
 
+$(PKG_BUILD_DIR)/uart16550.o: src/uart16550.c
+	$(TARGET_CC) $(CFLAGS) -c -o $@ $<
+
 $(STAGING_DIR)/bin/srec2bin: src/srec2bin.c
 	$(HOSTCC) -o $@ $<
 
@@ -41,11 +47,14 @@
 $(KDIR)/vmlinux.bin: $(KDIR)/vmlinux.lzma compile
 	$(TARGET_CROSS)ld -T $(PKG_BUILD_DIR)/zimage.script -r -b binary $< -o $(KDIR)/zimage.o
 	$(TARGET_CROSS)ld -static -G 0 --defsym kernel_entry=$(KERNEL_ENTRY) -T $(PKG_BUILD_DIR)/ld.script \
+		$(PKG_BUILD_DIR)/entry.o \
 		$(PKG_BUILD_DIR)/loader.o \
 		$(PKG_BUILD_DIR)/LzmaDecode.o \
+		$(PKG_BUILD_DIR)/uart16550.o \
 		$(KDIR)/zimage.o \
 		-o $(KDIR)/loader
 	$(TARGET_CROSS)objcopy -O srec $(KDIR)/loader $(KDIR)/ram_zimage.sre
+	$(TARGET_CROSS)objcopy -O binary $(KDIR)/loader $(BIN_DIR)/openwrt-$(BOARD)-$(KERNEL)-kernel.raw
 	$(STAGING_DIR)/bin/srec2bin $(KDIR)/ram_zimage.sre $@
 	
 $(BIN_DIR)/openwrt-$(BOARD)-$(KERNEL)-kernel.bin: $(KDIR)/vmlinux.bin 
@@ -84,7 +93,8 @@
 	rm -f $(BIN_DIR)/openwrt-$(BOARD)*
 
 prepare: $(PKG_BUILD_DIR) $(PKG_BUILD_DIR)/zimage.script $(PKG_BUILD_DIR)/ld.script
-compile: prepare $(PKG_BUILD_DIR)/loader.o $(PKG_BUILD_DIR)/LzmaDecode.o $(STAGING_DIR)/bin/srec2bin
+compile: prepare $(PKG_BUILD_DIR)/loader.o $(PKG_BUILD_DIR)/LzmaDecode.o $(PKG_BUILD_DIR)/uart16550.o $(STAGING_DIR)/bin/srec2bin
+compile: $(PKG_BUILD_DIR)/entry.o
 else
 clean:
 prepare:
Index: target/linux/image/ar7/src/loader.c
===================================================================
RCS file: /openwrt/openwrt/target/linux/image/ar7/src/loader.c,v
retrieving revision 1.2
diff -u -b -B -u -r1.2 loader.c
--- target/linux/image/ar7/src/loader.c	25 Jul 2005 09:34:53 -0000	1.2
+++ target/linux/image/ar7/src/loader.c	2 Sep 2005 18:44:58 -0000
@@ -15,21 +15,23 @@
 #include <linux/config.h>
 #include "gzip.h"
 #include "LzmaDecode.h"
+#include "loader.h"
+#include "uart16550.h"
 
 /* Function prototypes */
-unsigned char get_byte(void);
-int tikernelunzip(int,char *[], char *[]);
 static int tidecompress(uch *, uch *);
+static unsigned char get_byte(void);
+static void printf(const char *ptr);
+static void print_i(int i);
 
-void kernel_entry(int, char *[], char *[]);
-void (*ke)(int, char *[], char *[]); /* Gen reference to kernel function */
-void (*prnt)(unsigned int, char *);		/* Gen reference to Yamon print function */
-void printf(char *ptr);			/* Generate our own printf */
+extern void kernel_entry(int, char *[], char *[]);
+extern unsigned int workspace;
+extern unsigned char kernelimage[], kernelimage_end[];
 
-int tikernelunzip(int argc, char *argv[], char *arge[])
+static void (*ke)(int, char *[], char *[]); /* Gen reference to kernel function */
+
+int tikernelunzip2(int argc, char *argv[], char *arge[])
 {
-	extern unsigned int _ftext;
-	extern uch kernelimage[];
 	uch *in, *out;
 	int status;
 
@@ -54,8 +56,8 @@
 }
 
 #if 0
-char hex[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
-void print_i(int i)
+static const char hex[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+static void print_i(int i)
 {
 	int j;
 	char buf[11];
@@ -74,10 +76,8 @@
 }
 #endif
 
-int tidecompress(uch *indata, uch *outdata)
+static int tidecompress(uch *indata, uch *outdata)
 {
-	extern unsigned int workspace;
-	extern unsigned char kernelimage[], kernelimage_end[];
 	unsigned int i;  /* temp value */
 	unsigned int lc; /* literal context bits */
 	unsigned int lp; /* literal pos state bits */
@@ -120,15 +120,18 @@
 	return status;
 }
 
-
-void printf(char *ptr)
+static void printf(const char *ptr)
 {
-	unsigned int *tempptr = (unsigned int  *)0x90000534;
-	prnt = ( void (*)(unsigned int, char *)) *tempptr;
-	(*prnt)(0,ptr);
+	while (*ptr) {
+		if (*ptr == '\n') {
+			Uart16550Put('\r');
+		}
+		Uart16550Put(*ptr);
+		ptr++;
+	}
 }
 
-unsigned char get_byte()
+static unsigned char get_byte(void)
 {
 	unsigned char c;
 	
Index: target/linux/linux-2.4/ar7.mk
===================================================================
RCS file: /openwrt/openwrt/target/linux/linux-2.4/ar7.mk,v
retrieving revision 1.3
diff -u -b -B -u -r1.3 ar7.mk
--- target/linux/linux-2.4/ar7.mk	25 Jul 2005 19:25:14 -0000	1.3
+++ target/linux/linux-2.4/ar7.mk	2 Sep 2005 18:44:58 -0000
@@ -23,6 +23,9 @@
 	zcat $(DL_DIR)/$(ATM_FIRMWARE_FILE) | tar -C $(BUILD_DIR) $(TAR_OPTIONS) -
 	touch $@
 
+$(eval $(call KMOD_template,DIAG,diag,\
+	$(MODULES_DIR)/kernel/drivers/net/diag/diag.o \
+))
 
 $(eval $(call KMOD_template,SANGAM_ATM_A,sangam-atm-annex-a,\
 	$(MODULES_DIR)/kernel/drivers/atm/tiatm.o \
Index: target/linux/linux-2.4/config/ar7
===================================================================
RCS file: /openwrt/openwrt/target/linux/linux-2.4/config/ar7,v
retrieving revision 1.18
diff -u -b -B -u -r1.18 ar7
--- target/linux/linux-2.4/config/ar7	22 Aug 2005 12:20:23 -0000	1.18
+++ target/linux/linux-2.4/config/ar7	2 Sep 2005 18:44:58 -0000
@@ -845,6 +845,9 @@
 # CONFIG_JFFS2_BBC_LZSS is not set
 # CONFIG_CRAMFS is not set
 CONFIG_SQUASHFS=y
+CONFIG_SQUASHFS_EMBEDDED=y
+CONFIG_SQUASHFS_FRAGMENT_CACHE_SIZE=3
+# CONFIG_SQUASHFS_VMALLOC is not set
 CONFIG_TMPFS=y
 CONFIG_RAMFS=y
 CONFIG_ISO9660_FS=m
